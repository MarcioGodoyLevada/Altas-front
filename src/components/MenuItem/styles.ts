import styled, { css, DefaultTheme } from 'styled-components';
import { MenuItemProps } from '.';

const modifiers = {
  disabled: () => css`
    box-shadow: 0 0.4rem 1.2rem #00000029;
    opacity: 0.5;
    cursor: not-allowed;
  `,

  hoverFocus: () => css`
    :hover,
    :focus {
      background: ${(props) => props.theme.colors.gray50};
      border-radius: 0.5rem;

      p {
        color: ${(props) => props.theme.colors.primary700};
      }
    }
  `,

  expanded: (theme: DefaultTheme) => css`
    border-radius: 0.5rem;
    background: ${theme.colors.gray50};

    p {
      color: ${(props) => props.theme.colors.primary700};
    }
  `,
};

export const Wrapper = styled.li`
  display: flex;
  flex-direction: column;
  background: ${(props) => props.theme.colors.whiteFull};
  border-radius: 0.5rem;
  margin: 1.2rem;
`;

export const Item = styled.div<MenuItemProps>`
  ${({ disabled, expanded, expandable, theme, path }) => css`
    height: 5rem;
    padding: ${expandable ? '2.8rem' : '1rem'};
    display: flex;
    cursor: pointer;
    align-items: center;
    justify-content: ${expandable ? 'flex-start' : 'center'};

    p {
      margin: 0 1.6rem 0;
      color: ${theme.colors.gray500};
      font-family: ${theme.font.family};
      font-size: ${theme.font.sizes.medium};
    }

    div {
      display: flex;
      justify-content: space-between;
      width: 100%;
    }

    ${disabled && modifiers.disabled()}
    ${!disabled && modifiers.hoverFocus()}
    ${expanded && modifiers.expanded(theme)}
  `}
`;
