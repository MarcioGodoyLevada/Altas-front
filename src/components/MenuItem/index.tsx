import React, { useState } from 'react';
import * as S from './styles';

export type MenuItemProps = {
  icon?: React.ComponentType;
  disabled?: boolean;
  expanded?: boolean;
  expandable?: boolean;
  path?: string;
  title?: string;
  children?: React.ReactNode;
  onClick?: () => void;
};

const MenuItem = ({
  disabled,
  expanded = false,
  expandable,
  title = '',
  children,
  path,
  onClick,
}: MenuItemProps) => {
  const [expandedState, setExpandedState] = useState(expanded);

  function handleClick() {
    if (disabled) {
      return;
    }

    if (onClick) {
      onClick();
      return;
    }

    setExpandedState(!expandedState);
  }

  return (
    <S.Wrapper>
      <S.Item
        onClick={handleClick}
        expandable={expandable}
        data-testid="expandable-button"
      >
        <div>
          <p>{title}</p>
        </div>
      </S.Item>
    </S.Wrapper>
  );
};

export default MenuItem;
