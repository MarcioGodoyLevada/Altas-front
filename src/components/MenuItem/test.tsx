import { fireEvent, screen } from '@testing-library/react';
import { renderWithTheme } from '../../utils/tests/helpers';
import MenuItem from '.';

describe('Component: MenuItem', () => {
  it('should render correctly active menu item', () => {
    const { container } = renderWithTheme(
      <MenuItem
        key={1}
        expanded={true}
        title={'Click'}
        expandable={true}
        path={'/test'}
        onClick={() => {}}
      >
        <li key={1}>{'ativo'}</li>
        ))
      </MenuItem>
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(container).toHaveTextContent('Click');
    expect(container).toHaveTextContent('ativo');
  });

  it('should render correctly inactive menu item', () => {
    const { container } = renderWithTheme(
      <MenuItem
        key={1}
        expanded={false}
        title={'Click'}
        expandable={true}
        path={'/test'}
        onClick={() => {}}
      >
        <li key={1}>{'fechado'}</li>
        ))
      </MenuItem>
    );

    expect(container).toHaveTextContent('Click');
    expect(container.querySelector('svg')).toBeInTheDocument();
    expect(screen.queryByText('fechado')).not.toBeInTheDocument();
  });

  it('should render correctly menu item click', () => {
    renderWithTheme(
      <>
        <MenuItem
          key={1}
          expanded={true}
          title={'PrimaryClick'}
          expandable={true}
          path={'/test'}
          onClick={() => {}}
        >
          <li key={1}>{'PrimaryAtivo'}</li>
          ))
        </MenuItem>

        <MenuItem
          key={2}
          expanded={false}
          title={'SecondaryClick'}
          expandable={true}
          path={'/test'}
          onClick={() => {}}
        >
          <li key={2}>{'SecondaryAtivo'}</li>
          ))
        </MenuItem>
      </>
    );

    const PrimaryItem = screen.getByText('PrimaryClick');
    const SecondaryItem = screen.getByText('SecondaryClick');

    fireEvent.click(PrimaryItem);

    expect(PrimaryItem).toBeInTheDocument();
    expect(screen.getByText('PrimaryAtivo')).toBeInTheDocument();
    expect(SecondaryItem).toBeInTheDocument();
    expect(screen.queryByText('SecondaryAtivo')).not.toBeInTheDocument();
  });
});
