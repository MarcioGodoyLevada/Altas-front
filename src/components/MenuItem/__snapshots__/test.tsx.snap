// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: MenuItem should render correctly active menu item 1`] = `
.c3 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: #FFFFFF;
  border-radius: 0.5rem;
  margin: 1.2rem;
}

.c1 {
  height: 5rem;
  padding: 2.8rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  border-radius: 0.5rem;
  background: #F4F4F4;
}

.c1 p {
  margin: 0 1.6rem 0;
  color: #919191;
  font-family: Lato,sans-serif;
  font-size: 1.6rem;
}

.c1 div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

.c1:hover,
.c1:focus {
  background: #F4F4F4;
  border-radius: 0.5rem;
}

.c1:hover p,
.c1:focus p {
  color: #2C2657;
}

.c1 p {
  color: #222F3E;
}

.c5 {
  color: #2C2657;
  font-family: Lato,sans-serif;
  font-size: 1.6rem;
}

.c4 {
  color: #222F3E;
  width: 1.6rem;
  height: 1.4rem;
}

.c2 {
  color: #00ACE4;
  width: 2.4rem;
}

<li
  class="c0"
>
  <div
    class="c1"
    data-testid="expandable-button"
  >
    <svg
      class="c2"
    />
    <div>
      <p>
        Click
      </p>
      <svg
        aria-hidden="true"
        class="c3 c4"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21.886 5.536A1.002 1.002 0 0 0 21 5H3a1.002 1.002 0 0 0-.822 1.569l9 13a.998.998 0 0 0 1.644 0l9-13a.998.998 0 0 0 .064-1.033zM12 17.243 4.908 7h14.184L12 17.243z"
        />
      </svg>
    </div>
  </div>
  <ul
    class="c5"
  >
    <li>
      ativo
    </li>
    ))
  </ul>
</li>
`;
